### PR TITLE
typescript: Migrate components dir to typescript.

### PR DESCRIPTION
**Pull request by vsvipul**
**What's this PR do?**
Migrate components dir to typescript.

Fixed all errors except one-
```
app/renderer/js/components/tab.ts:19:49 - error TS2339: Property 'template' does not exist on type 'Tab'.

19  	this.$el = this.generateNodeFromTemplate(this.template());
    	                                              ~~~~~~~~
```
There was no template function in Tab class.

**You have tested this PR on:**
  - [ ] Windows
  - [x] Linux/Ubuntu
  - [ ] macOS

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -194,6 +194,7 @@
           "quotes": 0,
           "unicorn/catch-error-name": 0,
           "unicorn/no-console-spaces": 0,
+          "unicorn/regex-shorthand": 0,
```
Comment (vsvipul) [inline package.json]: Added to avoid linting error on line 141 of webview.ts.

Comment (priyank-p) [reply package.json]: @vsvipul we actually want this rule for improved readability of regex. The rule is also fixable so we can automate this.

File: app/renderer/js/components/handle-external-link.ts [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
+import LinkUtil = require('../utils/link-util');
+import DomainUtil = require('../utils/domain-util');
+import ConfigUtil = require('../utils/config-util');
+
+const { shell, app } = remote;
 
 const dingSound = new Audio('../resources/sounds/ding.ogg');
 
-function handleExternalLink(event) {
+function handleExternalLink(this: any, event: any): void {
```
Comment (priyank-p) [inline app/renderer/js/components/handle-external-link.ts]: I would add a todo comment here, to remind us to figure out a way to passed correct type here; we probably need to use a generic.

File: app/renderer/js/components/tab.ts [hunk 3]
```diff
@@ -1,9 +1,12 @@
 'use strict';
 
-const BaseComponent = require(__dirname + '/../components/base.js');
+import BaseComponent = require('../components/base');
 
 class Tab extends BaseComponent {
-	constructor(props) {
+	props: any;
```
Comment (priyank-p) [inline app/renderer/js/components/tab.ts]: Should be `object`, and we should also add a todo for this.

File: app/renderer/js/components/tab.ts [hunk 4]
```diff
@@ -1,9 +1,12 @@
...
+	webview: any;
```
Comment (priyank-p) [inline app/renderer/js/components/tab.ts]: We should also figure out type of this. Shouldn't be `any`.

Comment (priyank-p) [reply app/renderer/js/components/tab.ts]: Ah, so this should be `WebView` class.

File: app/renderer/js/components/server-tab.ts [hunk 5]
```diff
@@ -17,24 +19,23 @@ class ServerTab extends Tab {
...
-		this.$badge = this.$el.getElementsByClassName('server-tab-badge')[0];
+		this.$badge = this.$el.querySelectorAll('.server-tab-badge')[0];
 	}
 
-	updateBadge(count) {
+	updateBadge(count: number): void {
 		if (count > 0) {
-			const formattedCount = count > 999 ? '1K+' : count;
-
+			const formattedCount: string = count > 999 ? '1K+' : count.toString();
```
Comment (priyank-p) [inline app/renderer/js/components/server-tab.ts]: We don't need to explicitly type annotate to string here.

File: app/renderer/js/components/functional-tab.ts [hunk 6]
```diff
@@ -1,9 +1,10 @@
 'use strict';
 
-const Tab = require(__dirname + '/../components/tab.js');
+import Tab = require('../components/tab');
 
 class FunctionalTab extends Tab {
-	template() {
+	$closeButton: any;
```
Comment (priyank-p) [inline app/renderer/js/components/functional-tab.ts]: This should be `Element` or `HTMLElement`?

File: app/renderer/js/components/functional-tab.ts [hunk 7]
```diff
@@ -1,9 +1,10 @@
 'use strict';
```
Comment (priyank-p) [inline app/renderer/js/components/functional-tab.ts]: Also you forgot to do `export = FunctionTab;` at the bottom.

File: app/renderer/js/components/webview.ts [hunk 8]
```diff
@@ -1,30 +1,38 @@
 'use strict';
 
-const path = require('path');
-const fs = require('fs');
+import * as path from 'path';
+import * as fs from 'fs';
```
Comment (priyank-p) [inline app/renderer/js/components/webview.ts]: Remove the `* as` since we don't need it there.

File: app/renderer/js/components/webview.ts [hunk 9]
```diff
@@ -1,30 +1,38 @@
...
 
-const BaseComponent = require(__dirname + '/../components/base.js');
-const handleExternalLink = require(__dirname + '/../components/handle-external-link.js');
+const { app, dialog } = remote;
 
 const shouldSilentWebview = ConfigUtil.getConfigItem('silent');
+
 class WebView extends BaseComponent {
-	constructor(props) {
+	props: any;
```
Comment (priyank-p) [inline app/renderer/js/components/webview.ts:22]: should be `object` instead of `any`.

File: app/renderer/js/components/webview.ts [hunk 10]
```diff
@@ -1,30 +1,38 @@
...
+	zoomFactor: number;
+	badgeCount: number;
+	loading: boolean;
+	customCSS: string;
+	$webviewsContainer: any ;
```
Comment (priyank-p) [inline app/renderer/js/components/webview.ts]: Hmm, shouldn't this be `Element` or `HTMLElement`?

File: package.json (hunk 1 above)
Comment (vsvipul) [reply package.json]: @priyank-p Great. I'll work on fixing this in webview then.

**Review by priyank-p — COMMENTED** [review]
@vsvipul left bunch of comments can you address those? About the error you are facing, I think that should be `this.super.template()`.

File: app/renderer/js/components/tab.ts (hunk 4 above)
Comment (priyank-p) [reply app/renderer/js/components/tab.ts]: Add a todo comment here to remind use that we need to annotate this to `WebView`.

File: app/renderer/js/components/webview.ts (hunk 10 above)
Comment (vsvipul) [reply app/renderer/js/components/webview.ts]: Should be ```DOMTokenList```

File: package.json (hunk 1 above)
Comment (vsvipul) [reply package.json]: Fixed this in the last commit. Please check if if is correct. Worked on regex for the first time.

File: app/renderer/js/components/tab.ts (hunk 4 above)
Comment (vsvipul) [reply app/renderer/js/components/tab.ts]: @priyank-p I have tried using WebView here. Please check and tell if it is the correct way or not.

File: app/renderer/js/components/tab.ts [hunk 11]
```diff
@@ -29,7 +31,7 @@ class Tab extends BaseComponent {
 	}
 
 	isLoading(): boolean {
-		return this.webview.isLoading;
+		return this.webview.$el.isLoading();
```
Comment (priyank-p) [inline app/renderer/js/components/tab.ts:29]: Can you explain why this change is correct? Is this a bug from before?

**Comment by priyank-p** [comment]
@vsvipul merged everything except the one commit for which I left a question above.

File: app/renderer/js/components/tab.ts (hunk 11 above)
Comment (vsvipul) [reply app/renderer/js/components/tab.ts:29]: @priyank-p Yes... ```isLoading``` is a property of ```<webview>``` type but here ```this.webview``` was of type ```WebView``` class which we created. ```$el``` is of type ```<webview>``` . If this is correct, then yes, this should be a bug from before.. @abhigyank and @akashnimare Can you please confirm this?

Comment (abhigyank) [reply app/renderer/js/components/tab.ts:29]: From documentation looks correct. Does this log correct value when called?

Comment (vsvipul) [reply app/renderer/js/components/tab.ts:29]: When i checked this, i coudn't exactly test out its working, but on console.logging previous code return ```undefined```, while this returns ```boolean``` values.

**Comment by priyank-p** [comment]
@vsvipul Merged thanks! Next you can either work on `preference` directory or the `resources/messages.js` which should be really quick. I am working on `notfication` directory and I should be done soon.

These were the [changes](https://github.com/zulip/zulip-desktop/compare/adfab3db946cf254f29009b51fbacbb5470c4ee7..dae26b6b3cb493148f38cb069bd0a12215fd8aca) is made for the whole PR before merging. Just minor things.

(Also last couple of PR have been really great!)
